File: MatlabBinaryFiles/readDifarData.m
function [data, error] = readDifarData(fid, fileInfo, data)
% reads binary data stored by the Difar module.
%
% Inputs:
%   fid = file identifier
%   fileInfo = structure holding the file header, module header, a handle
%   data = a structure containing the standard data
%
% Output:
%   data = structure containing data from a single object
%

% initialize variables
error=false;

try
    
    % read Difar specific data
    dataLength = fread(fid, 1, 'int32');
    if (dataLength==0)
        return;
    end
    
    if (fileInfo.moduleHeader.version <= 1)
        data.startSample = fread(fid, 1, 'int64');
    end

    data.clipStart = fread(fid, 1, 'int64');
    
    if (fileInfo.moduleHeader.version <= 1)
        data.channelMap = fread(fid, 1, 'int32');
    end
    
    data.displaySampleRate = fread(fid, 1, 'float');
    data.demuxedLength = fread(fid, 1, 'int32');
    
    if (fileInfo.moduleHeader.version <= 1)
        minFreq = fread(fid,1,'float');
        maxFreq = fread(fid,1,'float');
        data.freqLimits = [minFreq maxFreq];
    end
    
    data.amplitude = fread(fid, 1, 'float');
    data.gain = fread(fid, 1, 'float');
    data.selAngle = fread(fid, 1, 'float');
    data.selFreq = fread(fid, 1, 'float');
    data.speciesCode = readJavaUTFString(fid);

    if (fileInfo.moduleHeader.version >= 1)
        data.trackedGroup = readJavaUTFString(fid);
    end
    
    data.maxVal = fread(fid, 1, 'float');
    data.demuxData = fread(fid, [data.demuxedLength,3], 'int16') * data.maxVal / 32767;
    
    data.numMatches = fread(fid, 1, 'int16');
    if (data.numMatches > 0)
        data.latitude = fread(fid, 1, 'float');
        data.longitude = fread(fid, 1, 'float');
        
        if (fileInfo.moduleHeader.version >= 1)
            errorX = fread(fid,1,'float');
            errorY = fread(fid,1,'float');
            data.errors = [errorX errorY 0];
        end
        
        for i=1:data.numMatches-1
            data.matchChan(i) = fread(fid, 1, 'int16');
            data.matchTime(i) = fread(fid, 1, 'int64');
        end
    end
        
    
catch mError
    disp(['Error reading ' fileInfo.fileHeader.moduleType '  data object.  Data read:']);
    disp(data);
    disp(getReport(mError));
    error=true;
end

File: R/readDifarData.R
#' @title Read Difar Data
#' 
#' @description Reads binary data stored by the Difar Processing module.
#'   
#' @param fid binary file identifier
#' @param fileInfo structure holding the file header and module header
#' @param data a structure containing standard data
#' @param skipLarge a flag of whether or not to skip reading the waveform
#' 
#' @return a structure containing data from a single object, and a logical
#'   flag if an error has occurred
#' 
#' @author Taiki Sakai \email{taiki.sakai@noaa.gov}
#' 
readDifarData <- function(fid, fileInfo, data, skipLarge = FALSE) {
    error <- FALSE
    
    tryCatch({
        dataLength <- pamBinRead(fid, 'int32', n=1)
        if(dataLength==0) {
            return(list(data=data, error=error))
        }
        
        version <- fileInfo$moduleHeader$version
        
        if(version <= 1) {
            data$startSample <- pamBinRead(fid, 'int64', n=1)
        }
        
        data$clipStart <- pamBinRead(fid, 'int64', n=1)
        
        if(version <= 1) {
            data$channelMap <- pamBinRead(fid, 'int32', n=1)
        }
        
        data$displaySampleRate <- pamBinRead(fid, 'float', n=1)
        data$demuxedLength <- pamBinRead(fid, 'int32', n=1)
        
        if(version <= 1) {
            minFreq <- pamBinRead(fid, 'float', n=1)
            maxFreq <- pamBinRead(fid, 'float', n=1)
            data$freqLimits <- c(minFreq, maxFreq)
        }
        
        data$amplitude <- pamBinRead(fid, 'float', n=1)
        data$gain <- pamBinRead(fid, 'float', n=1)
        data$selAngle <- pamBinRead(fid, 'float', n=1)
        data$selFreq <- pamBinRead(fid, 'float', n=1)
        data$speciesCode <- readJavaUTFString(fid)$str
        
        if(version >= 1) {
            data$trackedGroup <- readJavaUTFString(fid)$str
        }
        
        data$maxVal <- pamBinRead(fid, 'float', n=1)
        if(skipLarge) {
            seek(fid, data$demuxedLength * 3 * 2, origin = 'current')
        } else {
        data$demuxData <- matrix(
            pamBinRead(fid, 'int16', n=data$demuxedLength * 3),
            nrow=data$demuxedLength, ncol=3) * data$maxVal / 32767
        }
        
        data$numMatches <- pamBinRead(fid, 'int16', n=1)
        if(data$numMatches > 0) {
            data$latitude <- pamBinRead(fid, 'float', n=1)
            data$longitude <- pamBinRead(fid, 'float', n=1)
            
            if(fileInfo$moduleHeader$version >= 1) {
                errorX <- pamBinRead(fid, 'float', n=1)
                errorY <- pamBinRead(fid, 'float', n=1)
                data$errors <- c(errorX, errorY, 0)
            }
            
            for(i in 1:(data$numMatches-1)) { # This seems wrong, why -1?
                data$matchChan[i] <- pamBinRead(fid, 'int16', n=1)
                data$matchTime[i] <- pamBinRead(fid, 'int64', n=1)
            }
        }
        
        return(list(data=data, error=error))
    # }, warning = function(w) {
    #     print(paste('Warning occurred: ', w))
    #     return(list(data=data, error=error))
    }, error = function(e) {
        print(paste('Error reading ', fileInfo$fileHeader$moduleType, ' data object. Data read:'))
        print(data)
        print(e)
        error <- TRUE
        return(list(data=data, error=error))
    })
}